### ISSUE TITLE
[BUG]: version 0.3.11 breaks behavior of decimal datatype

### ISSUE THREAD

**Issue by gerwinbrunner** [typeorm/typeorm#9616]
### Issue description

Datatype decimal get's broken in 0.3.11

### Expected Behavior

When using a Decimal column in Oracle and SQLLite it should be stored correspondingly.

```
  @Column({ type: "decimal", nullable: true })
  effectiveRate?: Maybe<number>;
```

```
create table LD_FX_RATES (
   ....
  "effectiveRate" decimal(23,15) null,
  ...
)
```

Version 0.3.10  is capable of storing values like:

0.03454 
0.23432
33.4522 
...

### Actual Behavior

After an upgrade to 0.3.11 storing the following values:

0.03454 
0.23432
33.4522 
...

will be stored as:

0
0
33

### Steps to reproduce

```
  @Column({ type: "decimal", nullable: true })
  effectiveRate?: Maybe<number>;
```

```
create table LD_FX_RATES (
   ....
  "effectiveRate" decimal(23,15) null,
  ...
)
```

use 0.3.10 and store and read a value.

Then use 0.3.11 and the behavior is as described above.

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |          |
| Node.js version     | 14.20.1  |
| Typescript version  | 4.8.4  |
| TypeORM version     | 0.3.11  |

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, but I don't know how to start. I would need guidance.

**Comment by iJhefe** [comment]
i think its happening after #9339 specifically in
https://github.com/typeorm/typeorm/blob/0d72317b3db3f93ce9d4c2f6f027c7486bbed08b/src/driver/postgres/PostgresDriver.ts#L804-L807

this is parsing any column from number to integer instead of just passing to number (float or integer)

**Comment by CollinCashio** [comment]
Hey @gerwinbrunner and @iJhefe ,

I am the one that added the line above in the PostgresDriver for TypeORM. There was a casting bug related to a test case that I was working on when adding the VirtualColumn decorator where number was not properly casting to int when being returned from the database.

If the data type as listed { type: "decimal" } is provided, this LOC should get ignored since type !== Number but equals "decimal". There was a similar discussion related to this and passing the decimal option under type resolved the issue.

When I pass the "decimal" type option, I am unable to replicate your issue @gerwinbrunner. I am providing my test case below. Let me know if I am missing anything.

```ts
const entity = await TimeSheet.save({
      hours: 123.24,
});

const pulledTimeSheet = await TimeSheet.findOne({
      where: {
          id: entity.id,
      },
});

/* Entity Definition */
@Entity({ name: "timesheets" })
export default class TimeSheet extends BaseEntity {
    @PrimaryGeneratedColumn()
    id: number;

    @Column({ type: "decimal", nullable: true })
    hours?: Maybe<number>;
}
```

**Comment by bblok11** [comment]
I can confirm that this is broken for MySQL as well

**Comment by bblok11** [comment]
Found the same code at line 691 of src/driver/mysql/MysqlDriver.ts. Numbers are always parsed to an int now, while a Number can be a decimal as well. This has major impact.

**Comment by rspieker** [comment]
This happens for multiple drivers:
- [AuroraMysqlDriver.ts:639](https://github.com/typeorm/typeorm/blob/master/src/driver/aurora-mysql/AuroraMysqlDriver.ts#L639)
- [MysqlDriver.ts:640](https://github.com/typeorm/typeorm/blob/master/src/driver/mysql/MysqlDriver.ts#L640)
- [MysqlDriver.ts:692](https://github.com/typeorm/typeorm/blob/master/src/driver/mysql/MysqlDriver.ts#L692)
- [OracleDriver.ts:558](https://github.com/typeorm/typeorm/blob/master/src/driver/oracle/OracleDriver.ts#L558)
- [PostgresDriver.ts:807](https://github.com/typeorm/typeorm/blob/master/src/driver/postgres/PostgresDriver.ts#L807)
- [SqlServerDriver.ts:579](https://github.com/typeorm/typeorm/blob/master/src/driver/sqlserver/SqlServerDriver.ts#L579)

The exact error is due to a difference in the type cast of the check versus the actual type cast for the value update.
```
value = !isNaN(+value) ? parseInt(value) : value
```
Mainly the difference between `+value` (which is the short hand notation of `Number(value)`) and `parseInt(value)` (parse the integer part of the string to a numer (= integer))

The fix would be to use the same casting method for both, e.g. `value = !isNaN(+value) ? +value : value`, and probably make it explicit for integers too (`parseInt(value, 10)`, here the radix 10 should be provided too, as it will [not automatically always be 10](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#description))

**Comment by jedsmith13** [comment]
Any update on this? Forcing Number to be an INT is not intuitive. 

I am using TypeORM on an existing database and this patch update has a lot of braking changes.

**Comment by pleerock** [comment]
@CollinCashio since there are multiple reports on the issue - is it possible if we revert the changes?

**Comment by CollinCashio** [comment]
> @CollinCashio since there are multiple reports on the issue - is it possible if we revert the changes?

Yes @pleerock , we can revert that change

Although, I would recommend updating the change to be a parseFloat instead of a parseInt since that is what most everyone is saying they have an issue with that (ie. It is not intuitive to cast a Number to an int - although this is how typeorm acts when creating a int data column migration)

This would also be preferred because it would continue to auto cast the typedef pulled down from the virtual column.

**Comment by gerwinbrunner** [comment]
@CollinCashio 
Hi we just tested with 0.3.16 and it seem that the behaviour is still present.
You did not implement any change based on the above discussion, right?

**Comment by netojose** [comment]
Testing on 0.3.15, and the problem exists (using Postgres)

**Comment by netojose** [comment]
I found a solution, you should to describe the type of your field
```ts
@Column({ type: 'decimal' })
value!: number
```

If you use `type: 'float'` also should to work

**Comment by pleerock** [comment]
There is so much confusion in this issue. People just keep posting messages about their problem without analyzing if this issue is related to their or not.

Maybe I missed something but I don't really see how @CollinCashio's code could break the `decimal`. Based on the current code parseInt happens only if column type is `Number`. 

Can somebody debug and find why it gets into the parseInt line of code and how type becomes Number out of decimal in postgres?

Or alternatively, find the last working version and first non-working and find the code difference which lead to such behavior.

**Comment by bblok11** [comment]
@pleerock That's correct, parseInt only happens when the column type is number, while in the previous version parseFloat was used. Without specifying the type in the Column decorator, the only information that TypeORM has about the type is that it's a number. In Javascript a number can contain both a decimal and an integer, so parseFloat should be used to support both cases.

I could be that if you use TypeORM to generate your database, a number type is translated to an INT column instead of a FLOAT or DECIMAL. Personally I don't use this however, I create the database myself and just use TypeORM as ORM. In that case, TypeORM should recognize the database column as a FLOAT or DECIMAL and use parseFloat instead, or just use parseFloat everytime a column with type number is found.

**Comment by Anjaa84** [comment]
@bblok11  this issue is still not fixed. Still we have it as parseInt instead of parseFloat
